src/Raven.Server/Documents/DocumentTombstoneCleaner.cs

```diff
@@ -33,23 +35,35 @@ public void Initialize()
 
         public void Subscribe(IDocumentTombstoneAware subscription)
         {
-            lock (_locker)
+            _locker.Wait();
+
+            try
             {
                 _subscriptions.Add(subscription);
             }
+            finally
+            {
+                _locker.Release();
+            }
         }
 
         public void Unsubscribe(IDocumentTombstoneAware subscription)
         {
-            lock (_locker)
+            _locker.Wait();
+
+            try
             {
                 _subscriptions.Remove(subscription);
             }
+            finally
+            {
+                _locker.Release();
+            }
         }
 
-        internal bool ExecuteCleanup()
+        internal async Task<bool> ExecuteCleanup()
         {
-            if (Monitor.TryEnter(_locker) == false)
+            if (await _locker.WaitAsync(0) == false)
```

Comment:
Bad idea, this can be held up for a long time, during which you are holding up all the work that someone might do because they are trying to subscribe / unsubscribe.

Better to do the fast subscription work we need here to get the last observed stats, release the lock and then execute in the merger.